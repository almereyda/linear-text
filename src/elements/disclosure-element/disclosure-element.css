.disclosure-element__summary {
  /* Hide the drop-down triangle indicator. */
  list-style: none;
}

/* Hack: copied from button:active to avoid flicker when changing active
   states. */
.disclosure-element[open] > .disclosure-element__summary {
  border-style: inset;
}
