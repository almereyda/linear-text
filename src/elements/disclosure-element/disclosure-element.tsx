import {ReactNode, SyntheticEvent, useState} from 'react'

import {useCallback, useMemo} from 'react'

import './disclosure-element.css'

export type DisclosureElementProps = Readonly<{
  children: ReactNode
  'data-testid'?: string
  onToggle?(open: boolean): void
  open?: boolean
  summary: ReactNode
  title: string
}>

export function DisclosureElement({
  children,
  'data-testid': testID,
  onToggle,
  open: initOpen = false,
  summary,
  title
}: DisclosureElementProps): JSX.Element {
  const [open, setOpen] = useState(initOpen)

  const onToggleCb = useCallback(
    (ev: SyntheticEvent<HTMLDetailsElement>) => {
      setOpen(ev.currentTarget.open)
      onToggle?.(ev.currentTarget.open)
    },
    [onToggle, setOpen]
  )

  const summaryClassName = useMemo(
    () =>
      `disclosure-element__summary buttonish-element ${
        open ? 'buttonish-element--active' : ''
      }`,
    [open]
  )

  return (
    <details
      className='disclosure-element'
      data-testid={testID}
      onToggle={onToggleCb}
      open={open}
    >
      <summary className={summaryClassName} data-title={title} role='button'>
        {summary}
      </summary>
      {children}
    </details>
  )
}
