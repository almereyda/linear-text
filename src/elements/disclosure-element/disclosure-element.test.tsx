import {DisclosureElement} from './disclosure-element'
import {fireEvent, render, screen} from '@testing-library/react'

test('Matches snapshot.', () => {
  render(
    <DisclosureElement data-testid='disclosure' summary='Summary' title='Title'>
      Children
    </DisclosureElement>
  )
  const disclosure = screen.getByTestId('disclosure')
  expect(disclosure).toMatchSnapshot()
})

test('Renders summary and children.', () => {
  render(
    <DisclosureElement
      data-testid='disclosure'
      open
      summary='Summary'
      title='Title'
    >
      Children
    </DisclosureElement>
  )
  const disclosure = screen.getByTestId('disclosure')
  expect(disclosure).toHaveTextContent('SummaryChildren')
})

test('Reports toggle.', () => {
  const onToggle = jest.fn()
  render(
    <DisclosureElement
      data-testid='disclosure'
      onToggle={onToggle}
      summary='Summary'
      title='Title'
    >
      Children
    </DisclosureElement>
  )

  const disclosure = screen.getByTestId('disclosure')
  fireEvent(disclosure, new Event('toggle'))
  expect(onToggle).toHaveBeenCalledTimes(1)
})
