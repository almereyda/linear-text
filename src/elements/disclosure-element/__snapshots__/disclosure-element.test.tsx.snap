// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot. 1`] = `
<details
  class="disclosure-element"
  data-testid="disclosure"
>
  <summary
    class="disclosure-element__summary buttonish-element "
    data-title="Title"
    role="button"
  >
    Summary
  </summary>
  Children
</details>
`;
